### PR TITLE
Refactor http client, support chunked transfer.

### LINKED ISSUE #34: HTTP error on receiving chunked package.
## Bug description

HTTP client gets `Socket receive error` when handling header with `Transfer-Encoding: chunked`.

## Version

1.0.0-beta

## IOTA network

Which node are you connected to and which IOTA network is it in?

Hornet node.

## Hardware specification

Ubuntu laptop

## Steps To reproduce the bug

Explain how the maintainer can reproduce the bug.

1. connect to Hornet node
2. call `iota_client_get_trytes` API

## Expected behaviour

Get the response without errors.

## Errors

`cclient_core:    ERR: [iota_client_get_trytes:35] Socket receive error`

### PR DESCRIPTION
**Pull request by oopsmonk**
# Description of change

fixes #34 
HTTP client refactoring and support chunked transfer.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

local test via Hornet and IRI nodes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests using ginkgo that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: cclient/http/http.c [hunk 1]
```diff
@@ -33,180 +78,438 @@ static char const* header_template =
...
+  client->parser_settings->on_status = parser_on_status;
+  client->parser_settings->on_header_field = parser_on_header_field;
+  client->parser_settings->on_header_value = parser_on_header_value;
+  client->parser_settings->on_headers_complete = parser_on_headers_complete;
+  client->parser_settings->on_body = parser_on_body;
+  client->parser_settings->on_message_complete = parser_on_message_complete;
+  client->parser_settings->on_chunk_header = parser_on_chunk_header;
+  client->parser_settings->on_chunk_complete = parser_on_chunk_complete;
+  http_parser_init(client->parser, HTTP_RESPONSE);
+  client->parser->data = client;
```
Comment (wusyong) [inline cclient/http/http.c:341]: Sorry, I'm a little confused on how this works. Could you elaborate it?

File: cclient/api/examples/example_get_transactions_to_approve.c [hunk 2]
```diff
@@ -13,7 +13,7 @@
 void example_get_transactions_to_approve(iota_client_service_t *s) {
   printf("\n[%s]\n", __FUNCTION__);
   retcode_t ret = RC_ERROR;
-  flex_trit_t reference[FLEX_TRIT_SIZE_243];
+  // flex_trit_t reference[FLEX_TRIT_SIZE_243];
```
Comment (wusyong) [inline cclient/api/examples/example_get_transactions_to_approve.c:16]: Shouldn't this comment (and some others below) just be removed?

File: cclient/http/http.c (hunk 1 above)
Comment (oopsmonk) [reply cclient/http/http.c:341]: keeps the client object in the http_parser that we can use it in the http parser. 
see the parser callbacks like `parser_on_headers_complete`.

File: cclient/api/examples/example_get_transactions_to_approve.c (hunk 2 above)
Comment (oopsmonk) [reply cclient/api/examples/example_get_transactions_to_approve.c:16]: hmm...let me add a macro on it.

**Review by wusyong — APPROVED** [review]

**Pull request merged**